pkg/collector/corechecks/cluster/kubernetes_state_aggregators.go

```diff
@@ -0,0 +1,109 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021-present Datadog, Inc.
+
+// +build kubeapiserver
+
+package cluster
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/aggregator"
+	ksmstore "github.com/DataDog/datadog-agent/pkg/kubestatemetrics/store"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type metricAggregator interface {
+	accumulate(ksmstore.DDMetric)
+	flush(aggregator.Sender, *KSMCheck, *labelJoiner)
+}
+
+// maxNumberOfAllowedLabels contains the maximum number of labels that can be used to aggregate metrics.
+// The only reason why there is a maximum is because the `accumulator` map is indexed on the label values
+// and GO accepts arrays as valid map key type, but not slices.
+// This hard coded limit is fine because the metrics to aggregate and the label list to use are hardcoded
+// in the code and cannot be arbitrarily set by the end-user.
+const maxNumberOfAllowedLabels = 3
+
+type counterAggregator struct {
+	metricName    string
+	allowedLabels []string
+
+	accumulator map[[maxNumberOfAllowedLabels]string]float64
+}
+
+func newCounterAggregator(metricName string, allowedLabels []string) metricAggregator {
+	if len(allowedLabels) > maxNumberOfAllowedLabels {
+		log.Error("BUG in KSM metric aggregator")
+		return nil
+	}
+
+	return &counterAggregator{
+		metricName:    metricName,
+		allowedLabels: allowedLabels,
+		accumulator:   make(map[[maxNumberOfAllowedLabels]string]float64),
+	}
+}
+
+func (a *counterAggregator) accumulate(metric ksmstore.DDMetric) {
+	var labelValues [3]string
+
+	for i, allowedLabel := range a.allowedLabels {
+		if allowedLabel == "" {
+			break
+		}
+
+		labelValues[i] = metric.Labels[allowedLabel]
+	}
+
+	a.accumulator[labelValues] += metric.Val
+}
+
+func (a *counterAggregator) flush(sender aggregator.Sender, k *KSMCheck, labelJoiner *labelJoiner) {
+	for labelValues, count := range a.accumulator {
+
+		labels := make(map[string]string)
+		for i, allowedLabel := range a.allowedLabels {
+			if allowedLabel == "" {
+				break
+			}
+
+			labels[allowedLabel] = labelValues[i]
+		}
+
+		hostname, tags := k.hostnameAndTags(labels, labelJoiner)
+
+		sender.Gauge(ksmMetricPrefix+a.metricName, count, hostname, tags)
+	}
+
+	a.accumulator = make(map[[maxNumberOfAllowedLabels]string]float64)
+}
+
+var (
+	metricAggregators = map[string]metricAggregator{
```

Comment:
Any reason why don't we count statefulsets and daemonsets ?

Reply:
The main reason is that I took the list of objects to count and with which tag from [the legacy KSMv1 check: kubernetes_state.py#L90-L107](
https://github.com/DataDog/integrations-core/blob/7.26.0-rc.6/kubernetes_state/datadog_checks/kubernetes_state/kubernetes_state.py#L90-L107) and it wasn’t counting statefulsets and daemonsets.
But we can definitively add them !

Reply:
I think @DataDog/container-app is relying on these count metrics and might need them for sts and ds in the near future. Also for the sake of consistency we might want to add them now, wdyt?

Reply:
=> https://github.com/DataDog/datadog-agent/pull/7342/commits/009cb10199bf7eaba536a37f55c4389959adb7b0